### PR TITLE
Reply type decorator

### PR DESCRIPTION
**Pull request by aqk**
Note that part of this depends on calling the named protocol functions, and part of the checks are in `send_message` / `send_to_all`, `send_to_all_except`

When / before Almog lands https://github.com/Chia-Network/chia-blockchain/pull/8093/ I can submit a PR to his branch to join the two approaches (if possible).

### CODE REVIEW

File: chia/util/api_decorators.py [hunk 1]
```diff
@@ -59,3 +59,14 @@ def inner():
         return func
 
     return inner()
+
+
+def reply_type(type):
+    def wrap(func):
+        def inner():
+            setattr(func, "reply_type", type)
```
Comment (arvidn) [inline chia/util/api_decorators.py:67]: is this attribute used anywhere?

Comment (aqk) [reply chia/util/api_decorators.py:67]: Almog uses this in #8093 - in that patch, he is building the expected message return types into attributes on the sending functions.

Comment (arvidn) [reply chia/util/api_decorators.py:67]: I see

File: chia/protocols/protocol_state_machine.py [hunk 2]
```diff
@@ -58,3 +58,8 @@ def sent_message_response_ok(sent: ProtocolMessageTypes, received: Optional[Prot
             return False
 
     return True
+
+
+def main():
+    # Check this static invariant at import time
+    static_check_sent_message_response()
```
Comment (aqk) [inline chia/protocols/protocol_state_machine.py:65]: The need for this will go away when we remove `VALID_REPLY_MESSAGE_MAP`, and use only the `reply_type` decerator.

File: chia/protocols/protocol_timing.py [hunk 3]
```diff
@@ -0,0 +1,4 @@
+# These settings should not be end-user configurable
+INVALID_PROTOCOL_BAN_SECONDS = 10
```
Comment (aqk) [inline chia/protocols/protocol_timing.py:2]: I stuck these in here because I didn't want users fiddling with them as config settings.

File: chia/server/server.py [hunk 4]
```diff
@@ -615,7 +616,14 @@ async def send_to_all(self, messages: List[Message], node_type: NodeType):
                     if sent_message_response_ok(ProtocolMessageTypes(message.type), None):
                         await connection.send_message(message)
                     else:
-                        self.log.error(f"send_to_all not sending message expecting a response: {message.type}")
+                        # Internal peer protocol logic error
+                        # If we raise ProtocolError here exiting the loop will affect other peers
+                        self.log.error(f"send_to_all called with message expecting a response: {message.type}")
+                        await connection.close(
```
Comment (aqk) [inline chia/server/server.py:622]: These two calls to `connection.close` with `WSCloseCode.INTERNAL_ERROR` are because we do not expect to call `send_message` with peer messages that expect a reply.

See `chia/server/ws_connection.py` send_message

File: chia/server/ws_connection.py [hunk 5]
```diff
@@ -272,14 +279,20 @@ async def invoke(*args, **kwargs):
 
             msg: Message = Message(uint8(getattr(ProtocolMessageTypes, attr_name).value), None, args[0])
             request_start_t = time.time()
-            result = await self.create_request(msg, timeout)
+            result = await self.send_request(msg, timeout)
```
Comment (aqk) [inline chia/server/ws_connection.py:282]: unrelated name change - seems more clear

File: chia/util/errors.py [hunk 6]
```diff
@@ -7,7 +7,7 @@ class Err(Enum):
     DOES_NOT_EXTEND = -1
     BAD_HEADER_SIGNATURE = -2
     MISSING_FROM_STORAGE = -3
-    INVALID_PROTOCOL_MESSAGE = -4
+    INVALID_PROTOCOL_MESSAGE = -4  # We WILL ban for a protocol violation.
```
Comment (aqk) [inline chia/util/errors.py:10]: Note the comment on line 7:

`# temporary errors. Don't blacklist`

**Comment by aqk** [comment]
6580b3b bans peers for protocol violations, and also distinguishes a protocol error from our client (logic error) vs. a protocol reply error (peer bug or DOS)

File: chia/protocols/protocol_state_machine.py (hunk 2 above)
Comment (arvidn) [reply chia/protocols/protocol_state_machine.py:65]: does anyone call `main()`? The comment seem to suggest that it's called at import time, but I don't see it.
Did you mean to put this call in the global scope?

File: chia/server/server.py (hunk 4 above)
Comment (arvidn) [reply chia/server/server.py:622]: if we accidentally do this, and it's triggered by a test, we won't necessarily fail the test, because tests are allowed to print errors to the log

File: chia/protocols/protocol_state_machine.py [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+
+def static_check_sent_message_response() -> None:
+    """Check that allowed message data structures VALID_REPLY_MESSAGE_MAP and NO_REPLY_EXPECTED are consistent."""
+    # Reply and non-reply sets should not overlap: This check should be static
+    overlap = set(NO_REPLY_EXPECTED).intersection(set(VAILD_REPLY_MESSAGE_MAP.keys()))
+    if len(overlap) != 0:
+        raise AssertionError("Overlapping NO_REPLY_EXPECTED and VAILD_REPLY_MESSAGE_MAP values: {}")
+
+
+def sent_message_response_ok(sent: ProtocolMessageTypes, received: Optional[ProtocolMessageTypes]) -> bool:
```
Comment (arvidn) [inline chia/protocols/protocol_state_machine.py:46]: this name is a bit confusing to me. I would expect it to be called `message_response_ok()`, `validate_response()` or something along those lines.

It's checking whether the response is OK, right? the `sent`-prefix, indicating that it's a response to a message we *sent* seems redundant. Clearly we wouldn't receive a response to a message we *didn't* send.

**Review by arvidn — COMMENTED** [review]
is there a way we can test this?

File: chia/protocols/protocol_state_machine.py (hunk 7 above)
Comment (aqk) [reply chia/protocols/protocol_state_machine.py:46]: Right, renaming this

File: chia/server/server.py (hunk 4 above)
Comment (aqk) [reply chia/server/server.py:622]: @arvidn @almogdepaz What's a good solution here?

Should we save the fact that we had an error, and throw at the end of the loop?
Assert during tests?

Comment (aqk) [reply chia/server/server.py:622]: Ah, you know what, I can raise here, because the check is on the message type - it would affect all clients.

Comment (arvidn) [reply chia/server/server.py:619]: I think `raise` would be reasonable

**Review by arvidn — DISMISSED** [review]

**Review by wjblanke — APPROVED** [review]
aok
